Добавлена функциональность записи и чтения задач из файла

IT-Egor opened the pull request.

IT-Egor committed:
Создать новый класс для реализации TaskManager
Создать метод для сохранения в файл

IT-Egor committed:
Создать исключение ManagerSaveException

IT-Egor committed:
Написать реализацию класса FileBackedTaskManager

IT-Egor committed:
Создать метод toCSV() для всех типов задач для создания записи в CSV файле

IT-Egor committed:
Заменить Path в конструкторе на File

IT-Egor committed:
Реализовать чтение задач из файла

IT-Egor committed:
Изменить логику чтения из файла

IT-Egor committed:
Добавить методы создания FileBackedTaskManager в класс Managers

IT-Egor committed:
Вынести общие тесты для всех TaskManager в абстрактный класс

IT-Egor committed:
Написать тесты для FileBackedTaskManagerTest

IT-Egor committed:
Чистка кода

vlolad commented on src/taskmanager/exceptions/ManagerSaveException.java:
```diff
@@ -0,0 +1,7 @@
+    public ManagerSaveException(String message) {
+        super(message);
+    }
+}
```
У RuntimeException ещё есть конструктор с параметрами (String message, Throwable cause), во многих случаях его даже полезнее использовать для лучшего понимания ошибки

Для улучшения можно добавить, но не обязательно

vlolad commented on src/taskmanager/tasktypes/Task.java:
```diff
@@ -70,4 +70,16 @@ public String toString() {
                 '}';
     }
+
+    public String toCSVLine() {
```
Лучше сделать отдельный класс-конвертер, куда перенести методы toCSVLine() и fromCSVLine(). В целом будет достаточно и одного метода на конвертацию в/из CSV формат, который принимает родительский Task

Конвертировать таски можно в разные форматы, код класса может начать разбухать. Плюс многие поля повторяются, с помощью StringBuilder можно избежать дублирования. А все уникальные поля - добавлять через ветвления при необходимости

vlolad replied:
Ну и дополнительно можно в Task добавить поле-enum с типом задачи (TaskType, например, который содержит значения TASK, SUBTASK и EPIC). У нас на курсе рекомендуют такой подход вместо использования instanceof для определения того, какой класс у конкретного объекта задачи

Его же можно будет удобно переводить в строковое представление при конвертации в строку (и обратно). Но это по желанию, приму и с instanceof

vlolad commented on src/taskmanager/utility/Managers.java:
```diff
@@ -17,4 +20,12 @@ public static TaskManager getDefaultTaskManager() {
         return new InMemoryHistoryManager();
     }
+
+    public static TaskManager getFileTaskManager(File csvFile) {
```
По ТЗ новая реализация менеджера должна заменить getDefaultTaskManager()
При этом параметры у метода не нужны - вынеси дефолтный файл в качестве константы прямо в этом классе. Можно сделать вспомогательный метод с параметром, но getDefaultTaskManager() надо заменить

vlolad commented on src/taskmanager/utility/Managers.java:
```diff
@@ -17,4 +20,12 @@ public static TaskManager getDefaultTaskManager() {
+        return new FileBackedTaskManager(csvFile, Managers.getDefaultHistoryManager());
+    }
+
+    public static TaskManager loadTaskManagerFromFile(File file) {
```
Этот метод надо убрать, он дублирует метод loadFromFile(File file) из нового менеджера

vlolad commented on test/taskmanager/servise/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,122 @@
+                lines.add(reader.readLine());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
Ещё можно добавить assert на то, что исключение не выбрасывается. Но это для красоты)

vlolad commented on test/taskmanager/servise/InMemoryTaskManagerTest.java:
Старые тесты тоже нужно оставить :с
Они проверяют, что базовый функционал менеджера работает. Так как новый менеджер зависит от старого - это критично

vlolad commented on src/Main.java:
```diff
@@ -5,10 +5,18 @@
         TaskManager taskManager = Managers.getDefaultTaskManager();
+
+        Path path = Paths.get("/Users/egor.belousov/Desktop/Develop/Java/Courses/Projects/java-kanban/src/taskmanager/resourses/test.csv");
+        File csvFile = path.toFile();
```
Можно проще, сразу через конструктор File - new File("resources/task.csv")

vlolad commented on src/taskmanager/servise/impl/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,153 @@
+
+        } catch (IOException e) {
+            throw new ManagerSaveException("Ошибка чтения из файла");
+        }
```
По сохранению/восстановлению всё ок, только нужен конвертер (см. комментарий к Task)

vlolad reviewed (changes requested):
Привет!

Всё хорошо, хоть работает без сбоев
Что осталось поправить:
1) Вынести путь до файла в константу, и лучше сделать его относительно директории проекта (пришлось вручную менять на свой путь для проверки, не user-friendly)
2) Поправить комментарии в  классе Managers
3) Вынести конвертацию тасок в отдельный класс конвертера
4) Вернуть тесты старого менеджера

Остальное по мелочи, комментарии оставил. Если будут вопросы - пиши в Пачку
Хороших выходных!

IT-Egor committed:
Заменить конструктор в ManagerSaveException

IT-Egor committed:
Изменить реализацию метода getDefaultTaskManager() класса Managers, удалить лишние методы

IT-Egor committed:
Добавить задачам типы

IT-Egor committed:
Изменить логику чтения и записи в файл

Преобразование из файла в задачу и обратно теперь осуществляется через класс Converter

IT-Egor committed:
Чистка кода

IT-Egor committed:
Чистка кода

IT-Egor committed:
Чистка кода

vlolad commented on src/taskmanager/servise/impl/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,151 @@
+
+    public void save() {
+        try (Writer writer = new FileWriter(file)) {
+            //writer.write("test");
```
Тестовый коммент остался

vlolad commented on src/taskmanager/utility/Managers.java:
```diff
@@ -1,17 +1,21 @@
+import java.io.File;
+
 public class Managers {
+    public static final File CSV_FILE = new File("src/taskmanager/resources/test.csv");
```
Огонь

vlolad commented on src/taskmanager/tasktypes/Task.java:
```diff
@@ -1,19 +1,23 @@
         this.name = name;
         status = Status.NEW;
         id = 0;
+        type = Type.TASK;
```
Типы верно добавлены

vlolad commented on src/taskmanager/utility/Converter.java:
```diff
@@ -0,0 +1,41 @@
+import taskmanager.tasktypes.Subtask;
+import taskmanager.tasktypes.Task;
+
+public class Converter {
```
Получилось лучше, чем у меня в голове было. Жирный плюс за конвертер

Явное приведение типов ок, всего два раза и с обязательной проверкой типа задачи перед этим. Верная реализация, всё хорошо, так что можно не бояться использовать в умеренных количествах

vlolad commented on src/taskmanager/utility/Converter.java:
```diff
@@ -0,0 +1,41 @@
+            task = new Subtask(values[2], values[3]);
+            ((Subtask) task).setContainingEpicId(Integer.parseInt(values[5]));
+        } else {
+            throw new TaskTypeDoesNotExistException("Unknown task type: " + values[0]);
```
Можно сделать TaskTypeDoesNotExistException более общим классом ошибки и переименовать в ConvertationException или ConverterCSVException (например). Так как при конвертации могут встречаться и другие типы ошибок, то конкретные случаи вроде отсутствия типа задачи просто как сообщение передаются

Комментарий на улучшение, тут уже можно по желанию т.к. сильно усложнение конвертера не планируется далее в этом проекте. Просто на будущее с:

vlolad reviewed (approved):
Привет!

Конвертер огонь, всё остальное тоже огонь, можем двигаться к предпоследнему ТЗ
